sklearn/manifold/_locally_linear.py

```diff
@@ -43,23 +43,24 @@ def barycenter_weights(X, Z, reg=1e-3):
     See developers note for more information.
     """
     X = check_array(X, dtype=FLOAT_DTYPES)
-    Z = check_array(Z, dtype=FLOAT_DTYPES, allow_nd=True)
+    indices = check_array(indices, dtype=int)
 
-    n_samples, n_neighbors = X.shape[0], Z.shape[1]
+    n_samples, n_neighbors = X.shape[0], indices.shape[1]
     B = np.empty((n_samples, n_neighbors), dtype=X.dtype)
     v = np.ones(n_neighbors, dtype=X.dtype)
 
     # this might raise a LinalgError if G is singular and has trace
     # zero
-    for i, A in enumerate(Z.transpose(0, 2, 1)):
-        C = A.T - X[i]  # broadcasting
+    for i, ind in enumerate(indices):
+        A = X[ind]
+        C = A - X[i]  # broadcasting
```

Comment:
This is causing the tests to fail with:

```
        for i, ind in enumerate(indices):
>           A = X[ind]
E           IndexError: index 30 is out of bounds for axis 0 with size 20
```

- `Z` used to be `(n_samples, n_neighbors, n_features)`
- `Z.transpose(0, 2, 1)` used to be `(n_samples, n_features, n_neighbors)`
- `A` used to be `(n_features, n_neighbors)`
- `A.T` used to be `(n_neighbors, n_features)`
- `X[i]` used to be `(n_features,)`

so with broadcasting, `C` would be `(n_neighbors, n_features)` as well.

In the new code:

- `indices` is `(n_samples, n_neighbors)
- `ind` is `(n_neighbors,)`
- `A` is `(n_neighbors, n_features)`
- `X[i]` is still `(n_features,)`

so that should work... I am confused. Maybe use a debugger to check my assumptions.

Reply:
See: https://github.com/scikit-learn/scikit-learn/pull/17997/files#r460521313